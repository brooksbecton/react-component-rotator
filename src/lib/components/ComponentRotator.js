import React, { Component } from "react";
import PropTypes from "prop-types";

class ComponentRotator extends Component {
  constructor() {
    super();

    this.state = {
      activeIndex: 0,
      intervalId: undefined,
      timeoutId: undefined
    };
  }

  static defaultProps = {
    delay: 1000,
    startDelay: 0
  };

  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.string),
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]),
    delay: PropTypes.number,
    startDelay: PropTypes.number
  };

  componentDidUpdate(prevProps, prevState) {
    this.clearComponentInterval();
  }

  componentWillMount() {
    this.startTimer();
  }

  componentWillUnmount() {
    this.clearComponentInterval();
  }

  clearComponentInterval = () => {
    clearInterval(this.state.intervalId);
    clearInterval(this.state.timeoutId);
  };

  incrementIndex = () => {
    this.setState(prevState => ({
      activeIndex: (prevState.activeIndex + 1) % this.props.children.length
    }));
  };

  startInterval = () => {
    const intervalId = setInterval(this.incrementIndex, this.props.delay);
    this.setState({ intervalId });
  };

  startTimer = () => {
    this.startInterval();

    const timeoutId = setTimeout(() => {
      this.startInterval();
    }, this.props.startDelay);

    this.setState({ timeoutId });
  };

  render() {
    return <span>{this.props.children[this.state.activeIndex]}</span>;
  }
}

export default ComponentRotator;
