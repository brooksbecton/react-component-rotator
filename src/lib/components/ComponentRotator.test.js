import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React from "react";

import ComponentRotator from "./ComponentRotator";

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();

describe("ComponentRotator", () => {
  const defaultComponents = [<h2>Lie</h2>, <h2>Cheat</h2>, <h2>Steal</h2>];

  it("renders correctly", () => {
    const wrapper = shallow(<ComponentRotator children={defaultComponents} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("delays start of child by 'startDelay' prop", () => {
    const mockStartDelay = Math.random();

    shallow(
      <ComponentRotator
        children={defaultComponents}
        startDelay={mockStartDelay}
      />
    );
    expect(setTimeout).toHaveBeenLastCalledWith(
      expect.any(Function),
      mockStartDelay
    );
  });

  it("sets interval of child change by 'delay' prop", () => {
    const mockDelay = Math.random();

    shallow(
      <ComponentRotator children={defaultComponents} delay={mockDelay} />
    );

    expect(setInterval).toHaveBeenLastCalledWith(
      expect.any(Function),
      mockDelay
    );
  });

  it("clears intervals on unmount", () => {
    const mockIntervalId = Math.random();
    const mockTimeoutId = Math.random();

    const wrapper = shallow(<ComponentRotator children={defaultComponents} />);

    wrapper.setState({
      intervalId: mockIntervalId,
      timeoutId: mockTimeoutId
    });

    wrapper.unmount();

    expect(clearInterval).toHaveBeenCalledWith(mockIntervalId);
    expect(clearInterval).toHaveBeenCalledWith(mockTimeoutId);
  });

  it("clears intervals on prop change", () => {
    const mockIntervalId = Math.random();
    const mockTimeoutId = Math.random();

    const wrapper = shallow(<ComponentRotator children={defaultComponents} />);

    const moreComponents = [<h1>RTJ</h1>, ...defaultComponents];

    wrapper.setState({
      intervalId: mockIntervalId,
      timeoutId: mockTimeoutId
    });

    wrapper.setProps({ children: moreComponents });

    expect(clearInterval).toHaveBeenCalledWith(mockIntervalId);
    expect(clearInterval).toHaveBeenCalledWith(mockTimeoutId);
  });

  it("incrementIndex increments index", () => {
    const wrapper = shallow(<ComponentRotator children={defaultComponents} />);

    wrapper.instance().incrementIndex();

    expect(wrapper.instance().state.activeIndex).toBe(1);
  });

  it("incrementIndex keeps index within bounds", () => {
    const wrapper = shallow(<ComponentRotator children={defaultComponents} />);
    wrapper.instance().setState({ activeIndex: defaultComponents.length - 1 });
    wrapper.instance().incrementIndex();

    expect(wrapper.instance().state.activeIndex).toBe(0);
  });

  it("passes classes to the rendered HTML", () => {
    const testClassName= "byotiful";
    const wrapper = shallow(<ComponentRotator className={testClassName} children={defaultComponents} />);
    expect(wrapper.find("span").hasClass(testClassName)).toBeTruthy(); 
  });

});
