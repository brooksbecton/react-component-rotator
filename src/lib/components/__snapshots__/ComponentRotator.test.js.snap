// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentRotator renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ComponentRotator
    delay={1000}
    startDelay={0}
>
    <h2>
        Lie
    </h2>
    <h2>
        Cheat
    </h2>
    <h2>
        Steal
    </h2>
</ComponentRotator>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h2>
        Lie
</h2>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Lie",
      },
      "ref": null,
      "rendered": "Lie",
      "type": "h2",
    },
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h2>
          Lie
</h2>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Lie",
        },
        "ref": null,
        "rendered": "Lie",
        "type": "h2",
      },
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
